### ISSUE TITLE
Pipeline execute after update uses stale information when LDAP is enabled

### ISSUE THREAD

**Issue by costimuraru** [spinnaker/spinnaker#4941]
### Issue Summary:

Pipeline execution after a recent update will use stale information, when LDAP is enabled.

**Steps to reproduce:**
1. Enable LDAP in Spinnaker for authn/authz
2. Go in the UI and update an existing pipeline (eg. update a stage JSON)
3. Execute the pipeline (in a matter of a few seconds)

**Result:**
Execution uses old pipeline configuration (old json)

**Expected behavior:**
Execution should use the updated pipeline

### Feature Area (if this issue is UI/UX related, please tag `@spinnaker/ui-ux-team`):
Front50, Echo

### Description:

Whenever we update a pipeline (eg. from the UI), if we quickly run an Execute on that pipeline it will not fetch the latest updates (it will use the stale pipeline definition). It takes up to 30 seconds for the update to be picked up by Orca. If we disable LDAP (authn/authz), the problem does not manifest anymore.

From our investigation:
1. When trying to execute a pipeline from the UI, a call is made to echo
2. Echo will make a call to Front50 to fetch the latest pipeline configuration (see https://github.com/spinnaker/echo/blob/master/echo-core/src/main/java/com/netflix/spinnaker/echo/services/Front50Service.java#L20 )
3. Front50 always returns an empty list when LDAP is enabled (it returns the actual pipeline json if LDAP is disabled)
This can be tested via: `curl -v http://spin-front50.spinnaker:8080/pipelines/d996488f-9d92-4f07-bc85-6e56d1fbeef3/history?limit=1`
Returns: `[]`, even though pipeline exists.
4. Echo detects the empty list, logs a warning and **uses the cached pipeline configuration**.
```
c.n.s.e.pipelinetriggers.PipelineCache   : Got empty results back from front50's /pipelines/d996488f-9d92-4f07-bc85-6e56d1fbeef3/history?limit=1, falling back to cached=Pipeline(minhash-service-controller, Deploy minhash-service-controller, d996488f-9d92-4f07-bc85-6e56d1fbeef3)
```
5. Echo sends message to Orca **with the stale pipeline**.
6. Orca executes the pipeline with the old configuration.

The problem seems to be in front50: https://github.com/spinnaker/front50/blob/master/front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.groovy#L94-L95

Basically echo is not able to fetch data from front50 on the above endpoint, when LDAP is enabled.
Strangely enough, echo is not receiving a 403 (Forbidden), instead it receives an empty response (empty list).
As a quick test, I tried removing the `@PreAuthorize`/`@PostFilter` annotations on the getHistory() method and the problem is not manifesting anymore, even with LDAP enabled. 
I think Echo should be using anonymous auth when calling front50.

**Comment by ttomsu** [comment]
> Front50 always returns an empty list when LDAP is enabled (it returns the actual pipeline json if LDAP is disabled)
and 
> Strangely enough, echo is not receiving a 403 (Forbidden), instead it receives an empty response (empty list).

My hunch is that the empty list is being returned because the object is being filtered out of the returned `Collection` by the `@PostFilter`. While opening up Front50's endpoing to use a `?restricted=true` query param may fix your immediate issue, it's deliberately going around the auth controls that were put in place, which is not good. 

With LDAP turned on, try turning on Fiat debugging with this setting in `front50-local.yml`:

```
logging.level:
  com.netflix.spinnaker.fiat.shared.FiatPermissionEvaluator: DEBUG
```
and then call the Front50 endpoint. If calling from `curl`, you need to also pass the `X-SPINNAKER-USER` header, otherwise it will use the `anonymous` user (which I'm assuming doesnt' have access to this pipeline).

**Comment by costimuraru** [comment]
Thanks @ttomsu. Will enable logging and give it a try later today. If your assumption regarding PostFilter turns out to be true, what alternative would you recommend to fix the issue?

**Comment by ttomsu** [comment]
It could be a number of deeper issues - it could be the auth is working as expected, it could be Front50 has a caching issue, or the underlying storage of Front50 isn't working as fast as we'd like. What storage are you using with Front50?

**Comment by costimuraru** [comment]
> it could be Front50 has a caching issue, or the underlying storage of Front50 isn't working as fast as we'd like

Well, I think I excluded this one, because I've added a breakpoint in front50 [1], and it's able to retrieve the updated pipelines when calling getHistory. The problem is more likely, as you suggested due to the `@PostFilter`, which discards the elements post-factum (even though the DAO returns them just fine).

I tried with both S3 and SQL (AuroraDB) as the storage.

> While opening up Front50's endpoing to use a ?restricted=true query param may fix your immediate issue, it's deliberately going around the auth controls that were put in place, which is not good.

I see this behavior is already implemented for other Front50 endpoints, such as this one from the master branch: https://github.com/spinnaker/front50/blob/master/front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.groovy#L68


[1] https://github.com/spinnaker/front50/blob/master/front50-web/src/main/groovy/com/netflix/spinnaker/front50/controllers/PipelineController.groovy#L99

**Comment by ttomsu** [comment]
Were the Fiat debug logs any help? 

As for previous behavior, that was added because of other services (Clouddriver and Echo, IIRC) relied on caches of Front50 data to operate - basically there was no user those services could use to get everything. In your case, we do have a user, so we should go through the proper auth.

**Comment by ttomsu** [comment]
Haven't heard anything back in about a week. Closing for now - feel free to re-open if any new information comes up.
